SoundWire/SDCA: additional experiments with Functions

plbossart opened the pull request:
Tentative branch to show how simple it'd be to create subdevices for each SDCA function, using the auxiliary bus.

Tested on a MTL device with 2x SmartAmp, UAJ and HID with the following options added:

````
CONFIG_SND_SOC_SDCA=m
CONFIG_SND_SOC_SDCA_SMART_AMP=m
CONFIG_SND_SOC_SDCA_UAJ=m
CONFIG_SND_SOC_SDCA_SMART_MIC=m
CONFIG_SND_SOC_SDCA_HID=m
````

The code gets to a probe for all the drivers:
````
[   37.802023] acpi device:25: patch_sdca_function_type: found SDCA function UAJ (type 6)
[   37.802031] acpi device:26: patch_sdca_function_type: found SDCA function HID (type 10)
[   37.802347] acpi device:28: patch_sdca_function_type: found SDCA function SmartAmp (type 1)
[   37.802719] acpi device:2a: patch_sdca_function_type: found SDCA function SmartAmp (type 1)

[   37.812859] snd_soc_sdca_smart_amp snd_soc_sdca.SmartAmp.0: sdca_smart_amp_probe: done
[   37.812913] snd_soc_sdca_smart_amp snd_soc_sdca.SmartAmp.1: sdca_smart_amp_probe: done
[   37.819252] snd_soc_sdca_uaj snd_soc_sdca.UAJ.2: sdca_uaj_probe: done
[   37.819265] snd_soc_sdca_hid snd_soc_sdca.HID.3: sdca_hid_probe: done
````
````
root@fedora:~# ls -rt /sys/bus/auxiliary/devices/
snd_soc_sdca.SmartAmp.1  snd_soc_sdca.UAJ.2  soundwire_intel.link.2  soundwire_intel.link.0  snd_sof.hda-probes.0
snd_soc_sdca.SmartAmp.0  snd_soc_sdca.HID.3  soundwire_intel.link.1  snd_sof.msg_injector.0
````

````
root@fedora:~# ls -rt /sys/bus/soundwire/devices/sdw:*025d*/*sdca*
'/sys/bus/soundwire/devices/sdw:0:0:025d:0713:01/snd_soc_sdca.UAJ.2':
subsystem  driver  uevent  power

'/sys/bus/soundwire/devices/sdw:0:0:025d:0713:01/snd_soc_sdca.HID.3':
subsystem  driver  uevent  power

'/sys/bus/soundwire/devices/sdw:0:1:025d:1316:01/snd_soc_sdca.SmartAmp.0':
subsystem  driver  uevent  power

'/sys/bus/soundwire/devices/sdw:0:2:025d:1316:01/snd_soc_sdca.SmartAmp.1':
subsystem  driver  uevent  power
````

What we are missing is a callback to tell the functions that the init_io is complete on attachment. suspend-resume can be added as well.

@bardliao @ujfalusi @kv2019i let me know what you think.

plbossart committed:
ASoC: SDCA: add initial module

Add new module for SDCA (SoundWire Device Class for Audio) support.
For now just add a parser to identify the SDCA revision and the
function mask.

Note that the SDCA definitions and related MIPI DisCo properties are
defined only for ACPI platforms and extracted with _DSD helpers. There
is currently no support for Device Tree in the specification, the
'depends on ACPI' reflects this design limitation. This might change
in a future revision of the specification but for SDCA 1.0 ACPI is the
only supported type of platform firmware.

The SDCA library is defined with static inline fallbacks, which will
allow for unconditional addition of SDCA support in common parts of
the code.

The design follows a four-step process:

1) Basic information related to Functions is extracted from MIPI DisCo
tables and stored in the 'struct sdw_slave'. Devm_ based memory
allocation is not allowed at this point prior to a driver probe, so we only
store the function node, address and type.

2) When a codec driver probes, it will register subdevices for each
Function identified in phase 1)

3) a driver will probe for each subdevice and addition parsing/memory
allocation takes place at this level. devm_ based allocation is highly
encouraged to make error handling manageable.

4) Before the peripheral device becomes physically attached, register
access is not permitted and the regmaps are cache-only. When
peripheral device is enumerated, the bus level uses the
'update_status' notification; after optional device-level
initialization, the codec driver will notify each of the subdevices so
that they can start interacting with the hardware.

Note that the context extracted in 1) should be arguably be handled
completely in the codec driver probe. That would however make it
difficult to use the ACPI information for machine quirks, and
e.g. select different machine driver and topologies as done for the
RT712_VB handling later in the series. To make the implementation of
quirks simpler, this patchset extracts a minimal amount of context
(interface revision and number/type of Functions) before the codec
driver probe, and stores this context in the scope of the 'struct
sdw_slave'.

The SDCA library can also be used in a vendor-specific driver without
creating subdevices, e.g. to retrieve the 'initialization-table'
values to write platform-specific values as needed.

For more technical details, the SDCA specification is available for
public downloads at https://www.mipi.org/mipi-sdca-v1-0-download

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
soundwire: slave: lookup SDCA version and functions

Use SDCA helpers to get the basic information and store it in the
slave context. The information will be optionally be used in codec
drivers to register sub-devices for each Function.

When platforms are not based on ACPI the helpers do absolutely
nothing.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: SDCA: add quirk function for RT712_VB match

Add a generic match function for quirks, chances are we are going to
have lots of those...

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: Intel: soc-acpi: add is_device_rt712_vb() helper

Add a filter to skip the RT172 VB configuration if a SmartMic Function
is not found in the SDCA descriptors.

If the ACPI information is incorrect this can only be quirked further
with DMI information.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: SOF: Intel: hda: use machine_quirk() for SoundWire

Use the existing machine_quirk() callback to select an alternate
topology for RT712-VB devices.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: Intel: sof_sdw: add SmartMic DAI for RT712 VB

In theory the dailinks are created based on the number of endpoints
reported in ACPI match tables, so it should harmless to add a new
dailink: RT712 VA would not use it since it has only 2 endpoints.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: soc-acpi: introduce new 'machine check' callback

The existing machine_quirk() returns a pointer to a soc_acpi_mach
structure.
For SoundWire/SDCA support, we need a slightly different
functionality where a quirk function either validates or NACKs an
initial selection, based on additional firmware/DMI information.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC/soundwire: remove sdw_slave_extended_id

This structure is used to copy information from the 'sdw_slave'
structures, it's better to create a flexible array of 'sdw_slave'
pointers and directly access the information. This will also help
access additional information stored in the 'sdw_slave' structure,
such as an SDCA context.

This patch does not add new functionality, it only modified how the
information is retrieved.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
ASoC: SDCA: add function devices

Use the auxiliary bus to register/unregister subdevices for each
function. Each function will be handled with a separate driver,
matched using a name.

If a vendor wants to override a specific function driver, they could
use a custom name to match with a custom function driver.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart committed:
[HACK] rt1316/rt713: create subdevices for SDCA functions

root@fedora:/sys/bus/auxiliary/devices# ls -d *sdca*
snd_soc_sdca.HID.3  snd_soc_sdca.SmartAmp.0  snd_soc_sdca.SmartAmp.1  snd_soc_sdca.UAJ.2

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

bardliao commented on sound/soc/sdca/sdca_device.c:
```diff
@@ -22,3 +23,40 @@ void sdca_lookup_interface_revision(struct sdw_slave *slave)
+		if (id->part_id != 0x712 &&
+		    id->part_id != 0x713 &&
+		    id->part_id != 0x716 &&
+		    id->part_id == 0x717)
```
Is `==` a typo?

plbossart replied:
yep, will fix.

bardliao commented on sound/soc/sdca/sdca_device.c:
```diff
@@ -44,7 +44,7 @@ bool sdca_device_quirk_match(struct sdw_slave *slave, enum sdca_quirk quirk)
 		    id->part_id != 0x713 &&
 		    id->part_id != 0x716 &&
-		    id->part_id == 0x717)
+		    id->part_id != 0x717)
```
I believe the change belongs to `ASoC: SDCA: add quirk function for RT712_VB match`

plbossart replied:
right, the fixup was applied to the wrong patch. will fix.

plbossart committed:
ASoC: SDCA: add code to parse Function information

SDCA Function driver shall first parse the information from ACPI,
typically during their probe() callback.

All the data is devm_ allocated to avoid issues with error flows.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

bardliao commented on sound/soc/sdca/sdca_function_smart_amp.c:
```diff
@@ -0,0 +1,48 @@
+		dev_err(dev, "%s: %pfwP: probe failed: %d\n",
+			__func__, auxdev->dev.fwnode, ret);
+
+	return 0;
```
Should we return ret?

marc-hb commented:
Commit https://github.com/thesofproject/linux/pull/5010/commits/2772fe4c59b3ee25ceb4c5788b25a9664223b3d5 (merged into https://github.com/thesofproject/linux/commit/e7e81a5e790e) crashed `ba-lnlm-rvp-sdw-03` really bad and made it unavailable for CI for at least 7 hours (from Wed 2024-05-22 21:15:35 UTC to Thu 2024-05-23 03:18:11 UTC) because that devices (and other in BA) don't have  remote power control and can't be power-cycled by CI.

Can you please avoid too "risky" pull requests until they're fully equipped?

cc: @ssavati 

```
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: soundwire_bus:sdw_slave_set_frequency: rt711-sdca sdw:0:0:025d:0711:01: Configured bus base 1, scale 3, mclk 19200000, curr_freq 4800000
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: soundwire_bus:sdw_slave_set_frequency: rt715-sdca sdw:0:1:025d:0714:01: Configured bus base 1, scale 3, mclk 19200000, curr_freq 4800000
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: soundwire_bus:sdw_handle_slave_status: rt715-sdca sdw:0:1:025d:0714:01: signaling initialization completion for Slave 7
May 22 21:01:16 ba-lnlm-rvp-sdw-03 systemd[1]: Stopped target Sound Card.
May 22 21:01:16 ba-lnlm-rvp-sdw-03 acpid[382]: input device has been disconnected, fd 10
May 22 21:01:16 ba-lnlm-rvp-sdw-03 systemd[897]: Stopped target Sound Card.
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: snd_soc_rt1316_sdw:rt1316_io_init: rt1316-sdca sdw:0:3:025d:1316:01: rt1316_io_init hw_init complete
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: soundwire_bus:sdw_handle_slave_status: rt1316-sdca sdw:0:3:025d:1316:01: signaling initialization completion for Slave 1
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: snd_soc_rt1316_sdw:rt1316_io_init: rt1316-sdca sdw:0:2:025d:1316:01: rt1316_io_init hw_init complete
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: soundwire_bus:sdw_handle_slave_status: rt1316-sdca sdw:0:2:025d:1316:01: signaling initialization completion for Slave 1
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-lnl 0000:00:1f.3: ipc tx      : 0x44000000|0x3060004c: MOD_LARGE_CONFIG_SET [data size: 76]
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-lnl 0000:00:1f.3: ipc tx reply: 0x64000000|0x3060004c: MOD_LARGE_CONFIG_SET
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-lnl 0000:00:1f.3: ipc tx done : 0x44000000|0x3060004c: MOD_LARGE_CONFIG_SET [data size: 76]
May 22 21:01:16 ba-lnlm-rvp-sdw-03 kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-lnl 0000:00:1f.3: ipc tx      : 0x47000000|0x0: MOD_SET_DX [data size: 8]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-lnl 0000:00:1f.3: ipc tx reply: 0x67000000|0x0: MOD_SET_DX
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-lnl 0000:00:1f.3: ipc tx done : 0x47000000|0x0: MOD_SET_DX [data size: 8]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: snd_soc_rt711_sdca:rt711_sdca_calibration: rt711-sdca sdw:0:0:025d:0711:01: rt711_sdca_calibration calibration complete, ret=0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: snd_soc_rt711_sdca:rt711_sdca_jack_init: rt711-sdca sdw:0:0:025d:0711:01: in rt711_sdca_jack_init enable
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: snd_soc_rt711_sdca:rt711_sdca_io_init: rt711-sdca sdw:0:0:025d:0711:01: rt711_sdca_io_init hw_init complete
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: soundwire_bus:sdw_handle_slave_status: rt711-sdca sdw:0:0:025d:0711:01: signaling initialization completion for Slave 6
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: rt1316-sdca sdw:0:2:025d:1316:01: rt1316_sdw_remove: plb: before unregister_functions
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: BUG: kernel NULL pointer dereference, address: 0000000000000040
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: #PF: supervisor read access in kernel mode
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: #PF: error_code(0x0000) - not-present page
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: PGD 0 P4D 0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: Oops: 0000 [#1] PREEMPT SMP NOPTI
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: CPU: 2 PID: 20385 Comm: rmmod Not tainted 6.9.0-rc5-ge7e81a5e790e #pr5010
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: Hardware name: Intel Corporation Lunar Lake Client Platform/LNL-M LP5 RVP1, BIOS LNLMFWI1.R00.2470.D84.2312061937 12/06/2023
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: RIP: 0010:device_del+0x23/0x3d0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: Code: 90 90 90 90 90 90 90 f3 0f 1e fa 0f 1f 44 00 00 41 57 31 f6 41 56 41 55 4c 8d af 80 00 00 00 41 54 55 48 89 fd 53 48 83 ec 20 <4c> 8b 67 40 4c 89 ef 65 48 8b 04 25 28 00 00 00 48 89 44 24 18 31
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: RSP: 0018:ffff9729410b7ac8 EFLAGS: 00010282
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: RAX: 0000000000000000 RBX: 0000000000000001 RCX: 0000000000000000
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000000
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: RBP: 0000000000000000 R08: ffffffffa835e638 R09: 0000000100035a29
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: R10: ffffffffa8287260 R11: ffffffffa835e638 R12: ffff91ae45837000
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: R13: 0000000000000080 R14: 0000000000000080 R15: 0000000000000000
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: FS:  00007f4979861c40(0000) GS:ffff91b5a0c00000(0000) knlGS:0000000000000000
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: CR2: 0000000000000040 CR3: 0000000116252006 CR4: 0000000000f70ef0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: DR3: 0000000000000000 DR6: 00000000ffff07f0 DR7: 0000000000000400
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: PKRU: 55555554
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: Call Trace:
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  <TASK>
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? __die+0x24/0x70
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? page_fault_oops+0x15a/0x450
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? irq_work_queue+0x32/0x60
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? vprintk_emit+0x103/0x2f0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? local_clock+0x15/0x30
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? exc_page_fault+0x68/0x1e0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? asm_exc_page_fault+0x26/0x30
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? device_del+0x23/0x3d0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? _dev_info+0x70/0x90
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  sdca_dev_unregister_functions+0x36/0x60 [snd_soc_sdca]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  rt1316_sdw_remove+0x3e/0x60 [snd_soc_rt1316_sdw]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  sdw_drv_remove+0x54/0x80 [soundwire_bus]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  device_release_driver_internal+0x1a1/0x210
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  bus_remove_device+0xd4/0x140
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  device_del+0x151/0x3d0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? __mutex_unlock_slowpath+0x44/0x290
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  device_unregister+0x17/0x60
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  sdw_delete_slave+0xad/0xc0 [soundwire_bus]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? __pfx_sdw_delete_slave+0x10/0x10 [soundwire_bus]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  device_for_each_child+0x59/0xa0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  sdw_bus_master_delete+0x1e/0x60 [soundwire_bus]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  auxiliary_bus_remove+0x1c/0x30
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  device_release_driver_internal+0x1a1/0x210
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  bus_remove_device+0xd4/0x140
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  device_del+0x151/0x3d0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  sdw_intel_exit+0x89/0x110 [soundwire_intel]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  hda_dsp_remove+0x72/0x180 [snd_sof_intel_hda_generic]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  snd_sof_device_remove+0x119/0x180 [snd_sof]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  sof_pci_remove+0x1e/0x50 [snd_sof_pci]
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  pci_device_remove+0x3f/0xb0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  device_release_driver_internal+0x1a1/0x210
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  driver_detach+0x4b/0x90
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  bus_remove_driver+0x70/0xf0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  pci_unregister_driver+0x3f/0x90
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  __do_sys_delete_module+0x1a5/0x2a0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  ? __x64_sys_close+0x3c/0x80
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  do_syscall_64+0xa8/0x1b0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel:  entry_SYSCALL_64_after_hwframe+0x77/0x7f
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: RIP: 0033:0x7f4979989aeb
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: Code: 73 01 c3 48 8b 0d 45 33 0f 00 f7 d8 64 89 01 48 83 c8 ff c3 66 2e 0f 1f 84 00 00 00 00 00 90 f3 0f 1e fa b8 b0 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d 15 33 0f 00 f7 d8 64 89 01 48
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: RSP: 002b:00007ffead476c78 EFLAGS: 00000206 ORIG_RAX: 00000000000000b0
May 22 21:01:17 ba-lnlm-rvp-sdw-03 kernel: RAX: ffffffffffffffda RBX: 00005565dbe3e780 RCX: 00007f4979989aeb

```

plbossart commented:
> Can you please avoid too "risky" pull requests until all devices are fully equipped?

You would one know if a PR is risky or not?

Don't put a device in CI if it's not fully supported by the CI infrastructure...

marc-hb commented:
> You would one know if a PR is risky or not?

... when it says "experiments" in the name and causes "kernel NULL pointer dereferences", which I believe are fairly rare.

> Don't put a device in CI if it's not fully supported by the CI infrastructure...

It's better to have a device that must be manually rebooted from time to time compared to not having it at all. Manually rebooting it is a small (and temporary) price to pay for significantly more bandwidth. We're running really short in some areas.

plbossart commented:
> > You would one know if a PR is risky or not?
> 
> ... when it says "experiments" in the name and causes "kernel NULL pointer dereferences", which I believe are fairly rare.

Race conditions happen. I couldn't reproduce the problem on my own device with the scripts until I found a simpler recipe that shows it 100% of the time.

> > Don't put a device in CI if it's not fully supported by the CI infrastructure...
> 
> It's better to have a device that must be manually rebooted from time to time compared to not having it at all. Manually rebooting it is a small (and temporary) price to pay for significantly more bandwidth. We're running really short in some areas.

Those devices pay for themselves given the average engineer hourly rate.

marc-hb commented:
> Those devices pay for themselves given the average engineer hourly rate.

Come on, you know it's never that simple.

Anyway I think we keeping adding more hardware right now.

plbossart committed:
ASoC: SDCA: add helper to extract initialization_table

Existing vendor-specific codec drivers can retrieve the SDCA
initialization table for each Function. The use of the table value is
entirely up to the codec vendor, the values could be platform-specific
tuned values or used to patch the hardware.

Alternatively, the same table can be retrieved by a Function driver,
if enabled. The same low-level helper will be used in both cases.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
closing, will re-add a new branch
